test/e2e_node/image_pull_test.go

```diff
@@ -0,0 +1,286 @@
+				},
+			}
+			testpod2 = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "httpd2",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:            "httpd-new",
+						Image:           httpdNewImage,
+						ImagePullPolicy: v1.PullAlways,
+					}},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			}
+
+			ginkgo.By("cleanup images")
+			_ = RemoveImage(testpod.Spec.Containers[0].Image)
+			_ = RemoveImage(testpod2.Spec.Containers[0].Image)
+		})
+		ginkgo.AfterEach(func(ctx context.Context) {
+			ginkgo.By("cleanup pods")
+			if testpod != nil {
+				deletePodSyncByName(ctx, f, testpod.Name)
+			}
+			if testpod2 != nil {
+				deletePodSyncByName(ctx, f, testpod2.Name)
+			}
+		})
+
+		ginkgo.It("should pull immediately if no more than 5 pods", func(ctx context.Context) {
+			node := getNodeName(ctx, f)
+			testpod.Spec.NodeName = node
+			testpod2.Spec.NodeName = node
+
+			pod := e2epod.NewPodClient(f).Create(ctx, testpod)
+			pod2 := e2epod.NewPodClient(f).Create(ctx, testpod2)
+			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Failed", 30*time.Second, func(pod *v1.Pod) (bool, error) {
+				if pod.Status.Phase == v1.PodRunning || pod.Status.Phase == v1.PodFailed {
+					return true, nil
+				}
+				return false, nil
+			})
+			framework.ExpectNoError(err)
+			err = e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod2.Name, "Failed", 30*time.Second, func(pod *v1.Pod) (bool, error) {
+				if pod.Status.Phase == v1.PodRunning || pod.Status.Phase == v1.PodFailed {
+					return true, nil
+				}
+				return false, nil
+			})
+			framework.ExpectNoError(err)
+
+			events, err := f.ClientSet.CoreV1().Events(f.Namespace.Name).List(ctx, metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			var httpdPulled, httpdNewPulled pulledStruct
+			for _, event := range events.Items {
+				var err error
+				if event.Reason == kubeletevents.PulledImage {
+					if event.InvolvedObject.Name == testpod.Name {
+						httpdPulled, err = getDurationsFromPulledEventMsg(event.Message)
+						framework.ExpectNoError(err)
+					} else if event.InvolvedObject.Name == testpod2.Name {
+						httpdNewPulled, err = getDurationsFromPulledEventMsg(event.Message)
+						framework.ExpectNoError(err)
+					}
+				}
+			}
+
+			// as this is parallel image pulling, the waiting duration should be similar with the pulled duration.
+			// use 1.2 as a common ratio
+			if float32(httpdNewPulled.pulledIncludeWaitingDuration/time.Millisecond)/float32(httpdNewPulled.pulledDuration/time.Millisecond) > 1.2 {
+				framework.Failf("the pull duration including waiting %v should be similar with the pulled duration %v",
+					httpdNewPulled.pulledIncludeWaitingDuration, httpdNewPulled.pulledDuration)
+			}
+			if float32(httpdPulled.pulledIncludeWaitingDuration/time.Millisecond)/float32(httpdPulled.pulledDuration/time.Millisecond) > 1.2 {
```

Comment:
Is there more obvious way (without using hardcoded 1.2) to determine parallel pulling?

Reply:
No, I think(I cannot get other param to check this). 

The image pull happens in parallel. We don't have any metrics or flags to know the exactly start time of image pull and finished time. 

What we can get from kubelet are only these durations.
- 	pulledDuration               time.Duration
-	pulledIncludeWaitingDuration time.Duration

Reply:
Can we get a number of parallel pulls somehow?

Reply:
So we can monitor this value and if it's > 1 then pulling is parallel.
I'm worrying that current approach can cause flakes.

Reply:
> I prefer to add new test cases with more images/pods later. Currently, we don't have a single E2E test for this feature.

sounds good